### PR TITLE
Show Attendance Links Early

### PR DESCRIPTION
**Pull request by tessawiedmann**
We'd like the workshop attendance links to be accessible and show up on the workshop dashboard 48 hours before a given session starts, but we only actually let you use the link the day of the session. So going to the link before or after the day of the workshop will just take you to a message that says:

"Attendance for this workshop is currently closed. You can mark yourself as attended only on the day of your workshop. If you think you are seeing this message in error, please confirm your attendance link with your workshop organizer or email support@code.org for assistance."

Attendance links will be generated once a workshop admin presses "Start Workshop", and will only be shown 48 hours before and through midnight of the day of the session. The links will only work the day of the workshop. If you try to use the link outside of this time frame, relevant error messages are shown.

Test Case A
Create a 1-day workshop that begins on Monday
Start workshop on Friday
Don't see attendance link until Saturday

Test Case B
Create a 1-day workshop that begins on Monday
Start workshop on Saturday
See attendance link immediately after I start it

Test Case C
Create a 3-day workshop that begins on Monday
Start workshop on Saturday
See attendance link for day 1 immediately after I start it
On Sunday, see links for both day 1 and day 2.
On Monday, see links for all 3 days. Although, only day 1 works.
On Tuesday, day 1 link is hidden. See only day 2 and day 3. Only day 2 link works.

Test Case D
Create a 3-day workshop that begins on Monday
Start workshop on Tuesday
Attendance link for day 1 (Monday) doesn't work anymore on Tuesday

Cron job to remove stale session codes [here](https://github.com/code-dot-org/code-dot-org/pull/22085).

### CODE REVIEW

**Comment by tessawiedmann** [comment]
Both of these commits so far make it so that if an attendance link is visible, users can also mark themselves as attended. I need to update `open_for_attendance`, potentially update the tests I just committed, and update the Session controller tests.

**Comment by tessawiedmann** [comment]
Now I need to find all the places `open_for_attendance` is used and either change it to `show_links` when necessary or double check that `open_for_attendance` is correct. I will also need to update more tests.

File: apps/src/code-studio/pd/workshop_dashboard/workshop.jsx [hunk 1]
```diff
@@ -424,7 +424,8 @@ export default class Workshop extends React.Component {
       <div>
         <p>
           Every day of the workshop, your participants must visit the attendance URL to receive
-          professional development credit. The attendance URL will not show up until the day of the session.
+          professional development credit. The attendance URL will show up 48 hours early, but
+          participants can't mark themselves as attended until the day of the workshop.
```
Comment (tessawiedmann) [inline apps/src/code-studio/pd/workshop_dashboard/workshop.jsx]: @tanyaparker Is this updated wording okay? This is at the top of the "Take Attendance" section that shows up once you press "Start Workshop" 
image

File: dashboard/app/models/pd/session.rb [hunk 2]
```diff
@@ -78,11 +78,21 @@ def open_for_attendance?
   end
 
   def too_soon_for_attendance?
-    workshop.started_at.nil? || start - 12.hours > Time.zone.now
+    workshop.started_at.nil? || start.beginning_of_day > Time.zone.now
```
Comment (aoby) [inline dashboard/app/models/pd/session.rb]: Why this change? 

Note: the front ends are in UTC (which is 7 hours ahead of us in Seattle, 4 hours ahead of east coast), but the sessions start and end times are "local" times, not associated with a time zone. 

I'm not sure how specific we want this to be, but it might be more unpredictable using beginning and end of day rather than a set time difference.

File: dashboard/test/models/pd/session_test.rb [hunk 3]
```diff
@@ -100,9 +100,113 @@ class Pd::SessionTest < ActiveSupport::TestCase
...
-    assert session_ended.too_late_for_attendance?
+    refute session_ended.too_late_for_attendance?
+  end
+
+  test 'attendance link for one day workshop' do
+    #current day is Saturday
+    #workshop started Friday
+    workshop_a = create :pd_workshop, started_at: Time.now - 1.day
+    #session starts on Monday
+    session_a = create(:pd_session, workshop: workshop_a, start: Time.now + 2.days - 1.minute).tap(&:assign_code)
```
Comment (aoby) [inline dashboard/test/models/pd/session_test.rb]: nit: these comments are misleading: the current day is whatever day the test happens to be run. If you want it to be Friday, Saturday, etc then set that explicitly to match the comment but I'm not sure how much value that will add. Otherwise use relative words like `today`, `yesterday`, etc in the description.

What if we split this test case into two that each test a specific thing and are named as such. Then maybe we wouldn't even need the comments for readability.

Something like: 
 1. `workshop with first session tomorrow shows link but is not open for attendance`
 2. `workshop with first session in 3 days does not show link and is not open for attendance`

File: dashboard/test/models/pd/session_test.rb [hunk 4]
```diff
@@ -100,9 +100,113 @@ class Pd::SessionTest < ActiveSupport::TestCase
...
+  end
+
+  test 'attendance links for three day workshop' do
+    #current day is Saturday
+    #workshop started Saturday
+    workshop_c = create :pd_workshop, started_at: Time.now
+    #session 1 starts on Monday, 2 on Tuesday, 3 on Wednesday
+    session_c1a = create(:pd_session, workshop: workshop_c, start: Time.now + 2.days - 1.minute).tap(&:assign_code)
+    session_c2b = create(:pd_session, workshop: workshop_c, start: Time.now + 3.days - 1.minute).tap(&:assign_code)
+    session_c3c = create(:pd_session, workshop: workshop_c, start: Time.now + 4.days - 1.minute).tap(&:assign_code)
```
Comment (aoby) [inline dashboard/test/models/pd/session_test.rb]: These sessions don't need to be created individually. See the `num_sessions` and `sessions_from` attributes on the `pd_workshop` factory [here](https://github.com/code-dot-org/code-dot-org/blob/a4ad67204234dae0c9d13076071ba006a665bc8a/dashboard/test/factories/pd_factories.rb#L24-L26).

File: dashboard/test/models/pd/session_test.rb [hunk 5]
```diff
@@ -100,9 +100,113 @@ class Pd::SessionTest < ActiveSupport::TestCase
...
+    #current day is Saturday
+    #workshop started Saturday
+    workshop_b = create :pd_workshop, started_at: Time.now
+    #session starts on Monday
+    session_b = create(:pd_session, workshop: workshop_b, start: Time.now + 2.days - 1.minute).tap(&:assign_code)
+    refute session_b.open_for_attendance?
+    assert session_b.show_link?
+  end
+
+  test 'attendance links for three day workshop' do
```
Comment (aoby) [inline dashboard/test/models/pd/session_test.rb]: I'm not clear what this is testing. I wonder if we can scope this down more and split it into smaller more specific test cases with descriptive labels.

File: dashboard/app/models/pd/session.rb (hunk 2 above)
Comment (tessawiedmann) [reply dashboard/app/models/pd/session.rb]: I made this change so that when we say you can only mark yourself as attended on the day of the workshop, that it is actually true. If a workshop session starts after noon, you wouldn't be able to mark yourself as attended until 12 hours before the start date. Similarly, if the session starts early in the morning, you could mark yourself as attended before the day of the session. 

So what is `Time.zone.now` doing then? Is it not getting the "local" times?

File: dashboard/test/models/pd/session_test.rb (hunk 3 above)
Comment (tessawiedmann) [reply dashboard/test/models/pd/session_test.rb]: I was writing these test cases to match Tanya's test cases she sent me (in the description of this PR). I agree that using relative words is more correct, but thinking about it in terms of days of the week helped me visualize what was going on more. I think splitting it up would be helpful though, just wasn't sure if I could stray from the test cases I was given.

File: dashboard/test/models/pd/session_test.rb (hunk 4 above)
Comment (tessawiedmann) [reply dashboard/test/models/pd/session_test.rb]: That's super helpful thanks!

File: dashboard/app/models/pd/session.rb (hunk 2 above)
Comment (aoby) [reply dashboard/app/models/pd/session.rb]: `Time.zone.now` is the "local" time on the front end server, which is UTC. That will be the same regardless of where in the world the client viewing the result is. Remember all this Ruby / Rails code runs on the server and is always UTC.

The session start and end times are not relative to anything, they're just times that the workshop organizers fill in. Originally it was not used for anything automated. We have two main choices:

 1. Estimate like before. Instead of exactly beginning / end of day do a time span that roughly covers what we want across the US. 
 2. Be more sophisticated with session times and try to adjust based on the workshop location.

I still think using the actual beginning/end of day will be extra confusing and cause issues. For instance, midnight UTC is 5pm Seattle time, so the "next day" will start at 5pm here when we're likely to still have valid open workshops. Similarly the next day's workshop will start at 5pm the previous day.

File: dashboard/test/models/pd/session_test.rb (hunk 3 above)
Comment (aoby) [reply dashboard/test/models/pd/session_test.rb]: Hers are suggestions. They don't need to exactly match. We just need to be sure to test those scenarios.

File: dashboard/app/models/pd/session.rb (hunk 2 above)
Comment (tessawiedmann) [reply dashboard/app/models/pd/session.rb]: Okay that makes sense, I didn't realize it was all in UTC even when you use `.zone`.

I think estimating for beginning of day like it currently is will be fine. It did seem important that attendance ends around midnight of the day of the session, however. Could we set the cut off time to be 7am UTC? That way it's midnight for PST and 3am for EST?

File: dashboard/test/models/pd/session_test.rb (hunk 3 above)
Comment (tessawiedmann) [reply dashboard/test/models/pd/session_test.rb]: Gotcha. I'll split them up more and make them more descriptive.

File: dashboard/app/models/pd/session.rb (hunk 2 above)
Comment (aoby) [reply dashboard/app/models/pd/session.rb]: It's only because the time zone of our front end servers is UTC. That same code on your localhost will use your time zone, as expected (at least whatever your computer is configured to).

That estimate SGTM. @tanyaparker ?

Comment (tanyaparker) [reply dashboard/app/models/pd/session.rb]: Closing at midnight Pacific/3am Eastern seems fine to me. I'd rather err on keeping it open longer past midnight, than closing it too soon like 5pm.

File: apps/src/code-studio/pd/workshop_dashboard/workshop.jsx (hunk 1 above)
Comment (tanyaparker) [reply apps/src/code-studio/pd/workshop_dashboard/workshop.jsx]: Please change to:

There is a unique attendance URL for each day of your workshop. On each day of your workshop, your participants must visit that day's attendance URL to receive professional development credit. The attendance URL(s) will be shown below, 2 days in advance, for your convenience.

**Comment by tanyaparker** [comment]
Please take my test cases with a grain of salt. It's mostly a suggestion because that's how I was thinking I would test it manually when I review it. If you think there are better ways to structure it with less repetition or fewer specifics, that works for me! I would just want either you, the person reviewing your PR, or myself to make sure we try out my test cases before we ship.

File: dashboard/test/factories/pd_factories.rb [hunk 6]
```diff
@@ -30,16 +30,23 @@
       enrolled_unattending_users 0
       num_completed_surveys 0
       randomized_survey_answers false
+      assign_session_code false
+    end
+    trait :with_codes_assigned do
+      transient do
+        assign_session_code true
+      end
```
Comment (aoby) [inline dashboard/test/factories/pd_factories.rb]: I realize I suggested this, but I'm pretty sure we don't need to re-specify `transient` here since it's already defined as a transient attribute on line 33.

Try this as:
```ruby
trait :with_codes_assigned do
  assign_session_code true
end
```

Comment (tessawiedmann) [reply dashboard/test/factories/pd_factories.rb]: When we had it within the trait it wasn't in the bigger list of transient attributes, I think we just forgot to remove it when we added it in the bigger list. Fixed now- thanks for catching!

File: dashboard/test/models/pd/workshop_test.rb [hunk 7]
```diff
@@ -163,7 +163,7 @@ class Pd::WorkshopTest < ActiveSupport::TestCase
     @workshop.end!
     assert_equal 'Ended', @workshop.state
     assert_equal 'Ended', @workshop.state
-    assert @workshop.sessions.first.code.nil?
+    refute @workshop.sessions.first.code.nil?
```
Comment (aoby) [inline dashboard/test/models/pd/workshop_test.rb]: How about `assert_not_nil @workshop.sessions.first.code` ?

**Comment by clareconstantine** [comment]
[Show attendance link 48 hours in advance of the workshop until midnight the day of the workshop](https://trello.com/c/WPiSfjeT/530-show-attendance-link-48-hours-in-advance-of-the-workshop-until-midnight-the-day-of-the-workshop)
